[dsmr] Initial contribution, updated version

Hilbrand opened the pull request:
This is the updated version of the new DSMR binding for openHAB 2. It supersedes PR #2289 and closes issue #1608. In cooperation with @mvolaart I proceeded working on this binding to get a mergeable version and create a new PR that I can easily update.

For more details and configuration see the README.md file as part of this PR.

A version is available in the [Eclipse Market place](https://marketplace.eclipse.org/content/dutch-smart-meter-reading-dsmr-binding/). You should be able to install it via PaperUI (or disable/enable marketplace to refresh the binding list if you don't see the binding). 
The jar file (same as in the Eclipse Market place) is available at: https://github.com/Hilbrand/openhab2-addons/releases/download/DSMR/org.openhab.binding.dsmr-2.4.0-SNAPSHOT.jar

### Updating from an older version of this binding
The things and bridges have remained the same, but it now has some refresh configuration options. If you have an earlier version of this binding running, it might work by replacing the binding jar. However it now uses QuantityType and thus is very likely not compatible with openHAB versions below 2.4.0-SNAPSHOT. And even if you are already running that version due to this changes you might need to re-add the things if you added them via PaperUI.

### Some background
PR #2289 has been around for over a year. During that period, with some intervals, the code was reviewed and updated. The last half year nothing happened. Because I wanted this binding to work for me, I started to make changes to the code for myself. These changes became larger as I tried to get it to work for me to the point that they were nearing a stable code base. At that point the code had diverted significantly. To avoid others, especially @mvolaart, to do double work on the code I addressed this issue in the PR. Because of limited time @mvolaart gave me the ok to continue working on this binding and have it merged. 

### For reviewers
The original pr already has had quite some review comments. I believe all the issues mentioned in the pr that were open are addressed. But there are also some significant changes. This means a large part of the code needs to be reviewed again. In general all the xml files have mostly remained the same (except for the change to quantity type). And the parsing of the messages also remained the same. A good start might be to look at the last commit in this PR (this is the commit follows the commits from the original PR).  That commit also contains a list of functional and technical changes. Since I can't attach reviewers I'll tag people who have reviewed before or wanted to be informed: @martinvw @wborn @gedejong or feel free to add additional reviews. The more, the better as it can speedup merging this pr. I don't doubt there are some comments to make :smile:

### Notes on not yet added serial transport/serial-over-tcp
One feature was to add serial-over-tcp. I wanted to add this, but it turns out I found at least 3 ways serial-over-tcp can be done. The first is serial-over-tcp (RFC2217). This is planned to be supported out of the box in Smart Home. The second is having a virtual serial port that communicates over tcp. Probably this can achieved with the current implementation. The third would be to make a tcp connection and just send the serial data over tcp. Because of these choices I did prepare the code, but didn't proceed implementing support as it needs to be based on real use cases to see if there is extra work needed or if it will work out of the box.
Part of making it work would be to use the smart home serial transport module. This is not yet done as the module interface lacks some methods interfacing with the serial library that are used in this binding. It might be needed to add those methods to the module. But that would make this binding depend on the process of that support and I didn't want it introduce a blocking issue. But once these issues are worked out in the future, migrating would be fairly easy.

wborn commented:
Thanks for the update @Hilbrand! :+1:  I'll soon test the JAR and also have a look at the code.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDetector.java:
```diff
@@ -0,0 +1,86 @@
+                        || (prevDetectedMeter.getChannel().equals(meterDescriptor.getChannel())
+                                && meterType.requiredCosemObjects.length > prevDetectedMeter
+                                        .getMeterType().requiredCosemObjects.length)) {
+                    logger.info("New compatible meter: {}", meterDescriptor);
```
I'd rather only see my log filled with these messages when I specifically enable debug logging.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDiscoveryService.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    protected void startScan() {
+        logger.info("Start discovery on existing DSMR bridge.");
```
I'd rather only see my log filled with these messages when I specifically enable debug logging.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRMeterDiscoveryService.java:
```diff
@@ -0,0 +1,65 @@
+
+    @Override
+    protected synchronized void stopScan() {
+        logger.info("Stop discovery on existing DSMR bridge.");
```
I'd rather only see my log filled with these messages when I specifically enable debug logging.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,193 @@
+        scanning = false;
+        stopSerialPortScan();
+        super.stopScan();
+        logger.info("Stop DSMR discovery scan");
```
I'd rather only see my log filled with these messages when I specifically enable debug logging.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,193 @@
+     * Stops the serial port device.
+     */
+    private void stopSerialPortScan() {
+        logger.info("Stop discovery scan on port [{}].", currentScannedPortName);
```
I'd rather only see my log filled with these messages when I specifically enable debug logging.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,193 @@
+                    if (DSMRBindingConstants.DSMR_PORT_NAME.equals(portIdentifier.getCurrentOwner())) {
+                        logger.info("The port {} is owned by this binding. If no DSMR meters will be found it "
+                                + "might indicate the port is locked by an older instance of this binding. "
+                                + "Restart the system to unlock the port.", currentScannedPortName);
```
I'd rather only see my log filled with these messages when I specifically enable debug logging.

Hilbrand commented:
@wborn Thanks for the review. I've found the culprit, as in the area I kind of expected. It was the parsing of the unit related to the quantity type usages. This went wrong, but also was incorrectly handled when it went wrong resulting in incorrect detecting the slave energy meter. Still the way it detects meters is kind of shaky. It seems to work, but isn't precise with versions. The reason you see a version 5 gas meter, even if you have a version 4 meter. But I didn't touch most of that part.
I've also addressed you logging comments. These were related to the discovery process. In normal operations mode I believe I already had moved logging to debug. (If you want even more logging, try TRACE :smile: This will also log the raw data from the meter)

I've updated the Eclipse Marketplace url. So if you reinstall the binding you should have the fixes.

wborn commented:
Thanks for the quick fixes and tuning the log levels @Hilbrand!

I already made a local build and after some testing see that the Gas M3 meter is working properly again. :smiley:
The QuantityTypes also work well. So far the Things also keep staying online. With the original binding they continuously went offline/online. 

I'll keep testing it for a while and also have a more closer look at the code.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,299 @@
+     * Long running process that controls the DSMR device connection.
+     */
+    @Nullable
+    private DSMRDeviceRunnable dsmrDeviceRunnable;
```
Annotate the type with `@Nullable` instead:

```java
private @Nullable DSMRDeviceRunnable dsmrDeviceRunnable;
```

Please also fix all the other occurrences of this issue (also in other classes). :-)

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRDiscoveryService.java:
```diff
@@ -0,0 +1,88 @@
+        String thingId = "dsmr:" + meterType.name().toLowerCase() + ":"
+                + (meterDescriptor.getChannel() == DSMRMeterConstants.UNKNOWN_CHANNEL ? "default"
+                        : meterDescriptor.getChannel());
+        ThingUID thingUID = new ThingUID(thingId);
```
Can't you use `new ThingUID(thingTypeUID, id)` here instead?

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/connector/DSMRSerialConnector.java:
```diff
@@ -0,0 +1,282 @@
+                logger.trace("Opening port {}", serialPortName);
+                // Opening Operating System Serial Port
+                CommPortIdentifier portIdentifier = portManager.getIdentifier(serialPortName);
+                serialPort = portIdentifier.open(DSMRBindingConstants.DSMR_PORT_NAME,
```
To fix all these potential null pointer exceptions you need to assign `serialPort` to a local variable. Then do a null check and handle possible null values.

The reason for this is that the value of the `serialPort` property could be changed by another thread at any moment in time. So you will never know if `serialPort` is still not `null` after doing a null check. :-)

Can you also fix the other potential NPEs? Eclipse shows there are a total of 15 with 12 in this method.

Hilbrand replied:
O yes those null pointers warnings.... Maven shows me even more: 36! And these feel kind of sad. Doing an initialization of a field and in the next statement call a method on that field will result in a null pointer warning (although Eclipse doesn't warn me here, but maven does):
> dsmrDeviceThread = new Thread(dsmrDeviceRunnable);
> dsmrDeviceThread.start();   // <-- null pointer warning

Even wrapping code with a null check doesn't satisfy. I know theoretically they can be null, but given the working of the smart home framework it will not be called in a multi threading manner.  The tool-satisfying workaround is to introduce a local variable. Even if it theoretically does happen and I introduce a local variable. Do I really want to call start in case in between the initialization and calling start the field was assigned null? If I would introduce a local variable here I would have started a dangling thread that will never be closed since there is no reference. So this means to really fix it I would have to synchronize the call and the null assignment to that field to make sure it will work. That basically means all code must be made thread safe. But is that really needed?

Regarding the serialPort: it can never be null since all calls are wrapped in synchronized blocks,. Thus the warnings are all false-positives. The sad part: Introducing local variables, but removing the synchronized blocks would totally satisfy the tool, but also totally break the code...
**The current checking of @NonNullByDefault adds an implicit `ThreadSafe` to the class** and I'm not sure if that is really desired.

So I'm not sure what the best option is: Introduce local variables and possible cause really hard to find bugs because fields are potential assigned null and after that operations are performed on local variables or make all classes also threadsafe or simply remove the @NonNullByDefault?
Working around tooling always bothers me. I prefer to live with it's flaws.
So in the mean time I'll read them as to speak with my favorite line from Pirates of the caribbean:
> “The code is more what you’d call ‘guidelines’ than actual rules.” – Hector Barbossa

wborn replied:
Yes I agree the tooling could be a bit smarter at times. I think you're also allowed to use `@SuppressWarnings("null")` for suppressing real false positives.

wborn commented:
When would the following channels be updated?

* dsmr:device_v4:default:p1_text_code
* dsmr:device_v4:default:p1_text_string
* dsmr:device_v4:default:p1_timestamp

I haven't seen any state updates on these channels while they were linked.

The timestamp channel description suggests it should be updated for every new reading: "Timestamp of the last meter reading."

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRDiscoveryService.java:
```diff
@@ -67,10 +67,9 @@ public DSMRDiscoveryService() {
-                + (meterDescriptor.getChannel() == DSMRMeterConstants.UNKNOWN_CHANNEL ? "default"
-                        : meterDescriptor.getChannel());
-        ThingUID thingUID = new ThingUID(thingId);
+        String thingId = ":" + (meterDescriptor.getChannel() == DSMRMeterConstants.UNKNOWN_CHANNEL ? "default"
```
Thanks for the update! Though I think this ":" here should also be removed.

Hilbrand replied:
oeps.. yes the `:` is the default separator.

Hilbrand commented:
I have no idea when these channels are updated by the meter itself? If they don't change I assume they are not updated by the meter, because the fact that the device is added assumes they are correctly parsed. You could enable trace logging to see what the actual value is that is send by the meter.

martinvw commented on addons/binding/org.openhab.binding.dsmr/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,35 @@
+ org.openhab.binding.dsmr.handler,
+ org.osgi.framework,
+ org.slf4j,
+ tec.uom.se.quantity
```
Afaik we should normally not import this, @htreu? This should be hidden as an implementation detail IIRC

htreu replied:
But when my quick scan of the sources is right its not needed here, so just drop.

Hilbrand replied:
This was indeed not needed. I'll remove it.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,297 @@
+            dsmrDeviceRunnable = new DSMRDeviceRunnable(dsmrDevice, this);
+            dsmrDeviceThread = new Thread(dsmrDeviceRunnable);
+            dsmrDeviceThread.start();
+            watchdog = scheduler.scheduleWithFixedDelay(this::alive, receivedTimeoutNanos, receivedTimeoutNanos,
```
With such a short timing it might make sense to create your own scheduler and start both the `dsmrDeviceRunnable` and watch-dog in that, wdyt?

Because continuously (nano seconds scale sounds often) claiming a thread from the pool seems not good for the system as a whole.

Hilbrand replied:
The timing is in seconds. But the time is used to compare against `System.nanoTime()`. I'll used nanoseconds so I don't have to convert the system nano time to seconds each time the comparison is done. This comparison is used to determine the action in case of problems. It's a scaled recovery process.

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/handler/DSMRBridgeHandler.java:
```diff
@@ -0,0 +1,297 @@
+
+        logger.trace("Using configuration {}", deviceConfig);
+        updateStatus(ThingStatus.UNKNOWN);
+        receivedTimeoutNanos = TimeUnit.SECONDS.toNanos(deviceConfig.receivedTimeout);
```
Why do you / would you calculate on a nano scale if the input is seconds?

Hilbrand replied:
See previous comment

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRSerialAutoDevice.java:
```diff
@@ -0,0 +1,276 @@
+    /**
+     * Discovers the settings of the serial port.
+     */
+    DISCOVER_SETTINGS,
```
Minor I would expect that auto-formating would move this one tab to the right?

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRSerialAutoDevice.java:
```diff
@@ -0,0 +1,276 @@
+     * When switching baudrate ignore any errors received with the given time frame. Switching baudrate causes errors
+     * and should not be interpreted as reading errors.
+     */
+    private static final long SWITCHING_BAUDRATE_TIMEOUT_NANOS = TimeUnit.SECONDS.toNanos(5);
```
Why nano's

martinvw commented on addons/binding/pom.xml:
```diff
@@ -30,9 +30,11 @@
     <module>org.openhab.binding.chromecast</module>
     <module>org.openhab.binding.cm11a</module>
     <module>org.openhab.binding.coolmasternet</module>
+    <module>org.openhab.binding.denonmarantz</module>
```
:-D

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRDiscoveryService.java:
```diff
@@ -0,0 +1,85 @@
+    /**
+     * Timeout for Discovery timeout
+     */
+    private static final int DSMR_DISCOVERY_TIMEOUT_SECONDS = 120;
```
2 minutes?

Hilbrand replied:
I guess this can be shorter :smile: time between data can be 10 seconds, so detecting correct configuration could take 30 a 40 seconds. So rounded it up and multiplied it by 2. What would be a better timeout? 1 minute?

wborn replied:
Most bindings seem to settle for 10 seconds or 1 minute.
MqttServiceDiscoveryService even has 500 seconds. :frowning:

martinvw commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/P1TelegramParser.java:
```diff
@@ -0,0 +1,411 @@
+     */
+    private enum State {
+    /** Wait for the '/' character */
+    WAIT_FOR_START,
```
Formatting?

Hilbrand replied:
Same problem as other enum. For some reason this is how my eclipse formats it when saving the file.

martinvw commented:
@wborn can we expect additional comments from you? If not maybe we should just queue at @kaikreuzer because he already did a review round for the previous version so it should be an easy one this time 😄

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/p1telegram/P1TelegramParser.java:
```diff
@@ -0,0 +1,411 @@
+    /**
+     * Current crc value read.
+     */
+    private final StringBuilder crcValue = new StringBuilder();
```
Is it guaranteed that the StringBuilders in this class are used at any moment in time by at most 1 thread?

Hilbrand replied:
yes. the class is a state machine, it can only work with 1 thread otherwise the state would become a mesh.

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeter.java:
```diff
@@ -0,0 +1,86 @@
+    public DSMRMeter(DSMRMeterDescriptor meterDescriptor) {
+        this.meterDescriptor = meterDescriptor;
+
+        supportedIdentifiers = new ArrayList<>();
```
Can you move the initialization to the field and make it final?

wborn commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/meter/DSMRMeterConstants.java:
```diff
@@ -0,0 +1,26 @@
+ * Class containing constants that are applicable to the DSMRMeter
+ *
+ * @author M. Volaart - Initial contribution
+ */
```
Annotate with `@NonNullByDefault`.

Hilbrand replied:
Added `@NonNullByDefault`

wborn commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/i18n/dsmr_nl.properties:
```diff
@@ -0,0 +1,242 @@
+thing-type.config.dsmr.bridgesettings.enableAutoDetection.description = In het geval automatische seri�le poort niet werkt zet deze optie uit en stel de seri�le poort instelling handmatig in.
+thing-type.config.dsmr.bridgesettings.baudrate.label = Seri�le poort baudrate
+thing-type.config.dsmr.bridgesettings.baudrate.description = De seri�le poort baudrate (4800, 9600, 19200, 38400, 57600 of 115200).
+thing-type.config.dsmr.bridgesettings.databits.label= = Seri�le poort data bits
```
Remove one '=' here.

Hilbrand commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/DSMRBindingConstants.java:
```diff
@@ -0,0 +1,38 @@
+ *
+ * @author M. Volaart - Initial contribution
+ * @author Hilbrand Bouwkamp - Removed time constants
+ */
```
Added `@NonNullByDefault`

Hilbrand commented:
One general somewhat related comment. This binding uses the serial port. To get it to work some configuration might be needed (e.g. adding port to startup script, etc). I haven't included it in the readme. I saw at least 2 other bindings, serial and modbus describe these configuration. It would probably be a good idea to write some general documentation page about this as it's not binding specific and refer to it. What would be the best location (what entry point in the general documentation) to add such documentation? and I'll add it there and refer to that page from here.

kaikreuzer commented:
@Hilbrand I guess this is exactly what you are missing? https://github.com/openhab/openhab-docs/pull/755

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/DSMRSerialAutoDevice.java:
```diff
@@ -0,0 +1,276 @@
+        if (!telegram.getCosemObjects().isEmpty()) {
+            stopDiscover(DeviceState.NORMAL);
+            parentListener.handleTelegramReceived(telegram);
+            logger.info("Start receiving telegrams on port {} with settings: {}", dsmrConnector.getPortName(),
```
You can use "info" if this is a one time message at connection time, but here it seems that it can happen pretty much of the time?

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,34 @@
+Export-Package: org.openhab.binding.dsmr,
+ org.openhab.binding.dsmr.handler
+Import-Package: 
+ gnu.io,
```
Would be great if you could refactor the implementation to use the [new ESH Serial API](https://github.com/eclipse/smarthome/tree/master/bundles/io/org.eclipse.smarthome.io.transport.serial/src/main/java/org/eclipse/smarthome/io/transport/serial) (which is identical to gnu.io, just uses its own namespace) as this will also directly enable the serial-over-tcp support.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/README.md:
```diff
@@ -0,0 +1,297 @@
+# DSMR Binding
+
+## Introduction
```
remove this header - the intro is expected directly after the "binding" header

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java:
```diff
@@ -0,0 +1,157 @@
+    private synchronized void registerDiscoveryService(DSMRBridgeHandler bridgeHandler) {
+        DSMRMeterDiscoveryService discoveryService = new DSMRMeterDiscoveryService(bridgeHandler);
+
+        discoveryService.setLocaleProvider(localeProvider);
```
Hm, if I am not missing anything, neither the `localeProvider` nor the `i18nProvider` are actually used by the discovery service (and normally, there should also be no need for them). So you might want to remove the dependency to these services from the code.

Hilbrand replied:
The're used by the parent class `AbstractDiscoveryService` to set the label of a newly detected thing. These labels are i18n supported.

kaikreuzer replied:
Oh, thanks, I wasn't even aware of that :-)

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/device/connector/SerialPortManager.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Interface for a serial port manager.
+ *
+ * This interface is modeled after the smart home serial transport interface. Because that interface is not yet
```
What is missing in the ESH Serial API? I a not aware of any open issues that you might be waiting for.

Hilbrand replied:
I haven't created any open issues (yet) :smile: I did replace the class implementation with the ESH interface to test it and it misses some methods. I don't know if these method should be added or it it would require the usage to be changed a little. I didn't change much from the original implementation on how the methods of library are used from the original implementation so to not get off track too much I designed most of the code that the injection would be fairly easy and thus if it's as simple as adding the methods to the interface it should be fairly easy to add.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/discovery/DSMRBridgeDiscoveryService.java:
```diff
@@ -0,0 +1,213 @@
+/**
+ * This implements the discovery service for detecting new DSMR Meters.
+ *
+ * The service will iterate over the available serial ports and open the given serial port and wait for telegrams. If
```
Hm, we do not do such a port scanning in any other serial binding.
Isn't this mechanism pretty dangerous? I could imagine all kinds of connected serial devices that behave strangely if you open their port and get data from them. I would claim that a binding should not blindly connect to any serially attached device.

IF you claim that it is not dangerous: In such a case, we should introduce a SerialDiscoveryService in the framework and have bindings implement SerialDiscoveryParticipants, similar to UPnP, mDNS and Bluetooth.

In both cases my suggestion would be to remove this discovery service here.

kaikreuzer replied:
> it just waits for data to be received. So I don't think it's dangerous.

Well, some devices might become active in some (unintended) way by the fact that they are connected to and they can start sending data.
But ok, this discovery service is not automatically started (if I see it correctly), but only manually triggered by the user - so the risk is probably minimal (or at least avoidable by not triggering a discovery).

> I'm not sure if that can be generalized in a way that adds much value.

My thinking was that for background discovery, it would definitely make sense to centralize it - e.g. upon startup, the existing serial ports could be scanned for supported devices and also during runtime, if a device is plugged in and a new serial port becomes available, it would be cool to automatically populate the Inbox.
But if such logic were in the individual bindings, they would be concurrently trying to access the ports, which results in conflicts - a central SerialDiscoveryService could instead hand the port to each interested participant one after another. But that would be a topic for a different PR in ESH - if you find it interesting, feel free to create such a PR :-)

Hilbrand replied:
Yes the discovery process is only manually. That should reduce the risk. In most cases there are very few serial connections I guess (on average 2-3 com ports per system?) . So adding a new serial connected device probably requires some manual handwork to plugin the device in the openHAB server, and thus the use case for an auto discovery service is very low. It's not unlikely the device connected to the serial port does have multiple connections to other devices and there an auto discover might be useful (like rfxcom and this binding). But at that point it's not specific to the serial port anymore, but the protocol of the device and it's then on the port already assigned to the binding.
Also background discovery using the port scanning, might indeed be dangerous, as it does temporary lock the port each scan, and block the port for the actual device. I think as long as it's a manual one time process, while the user is looking should be fine. So I wouldn't recommend to doing background port scanning on a regular basis anyway.

kaikreuzer commented on addons/binding/org.openhab.binding.dsmr/ESH-INF/config/configuration_parameters.xml:
```diff
@@ -0,0 +1,88 @@
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/config-description/v1.0.0 http://eclipse.org/smarthome/schemas/config-description-1.0.0.xsd">
+
+	<config-description uri="thing-type:dsmr:bridgesettings">
+		<parameter-group name="serialPortAdvanced">
```
The usual mean in such situations is to mark the parameters as advanced, so they are by default not at all shown to the user. Wouldn't that be a good option here as well?
If you prefer keeping them as they are, it is ok for me as well, although it feels a bit unusual.

wborn commented:
It's finally merged! :heart_eyes: 
Thanks @mvolaart, @Hilbrand and all testers/reviewers!

martinvw commented:
@Hilbrand @wborn I just upgrade my environment to the latest build, I started it myself today so should include this changes but I cannot install the dsmr binding any ideas?

![image](https://user-images.githubusercontent.com/2625279/44870635-03df4780-ac91-11e8-9e58-7e6b54583031.png)

![image](https://user-images.githubusercontent.com/2625279/44870671-1e192580-ac91-11e8-9824-ddeef709fe22.png)

https://openhab.ci.cloudbees.com/job/openHAB-Distribution/1342/

Hilbrand commented:
The build you linked to is from 27 Augustus, the binding was merged on 28 Augustus. So I think it's not yet available.

kaikreuzer commented:
Yeah, but the addons are built in https://openhab.ci.cloudbees.com/job/openHAB2-Bundles/, which ran last on Aug 27...
So the DSMR is in the milestone build, but not in the latest snapshot 🙄 . I'll start the addons build right away.

martinvw commented:
So I triggered the wrong build 😅
